Sources/Baggage/Baggage.swift

```diff
@@ -0,0 +1,218 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Baggage Context open source project
+//
+// Copyright (c) 2020 Moritz Lang and the Swift Baggage Context project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+// ==== ----------------------------------------------------------------------------------------------------------------
+// MARK: Baggage
+
+/// A `Baggage` is a heterogeneous storage type with value semantics for keyed values in a type-safe fashion.
+///
+/// Its values are uniquely identified via `Baggage.Key`s (by type identity). These keys also dictate the type of
+/// value allowed for a specific key-value pair through their associated type `Value`.
+///
+/// ## Defining keys and accessing values
+/// Baggage keys are defined as types, most commonly case-less enums (as no actual instances are actually required)
+/// which conform to the `Baggage.Key` protocol:
+///
+///     private enum TestIDKey: Baggage.Key {
+///       typealias Value = String
+///     }
+///
+/// While defining a key, one should also immediately declare an extension on `Baggage`,
+/// to allow convenient and discoverable ways to interact with the baggage item, the extension should take the form of:
+///
+///     extension Baggage {
+///       var testID: String? {
+///         get {
+///           self[TestIDKey.self]
+///         } set {
+///           self[TestIDKey.self] = newValue
+///         }
+///       }
+///     }
+///
+/// For consistency, it is recommended to name key types with the `...Key` suffix (e.g. `SomethingKey`) and the property
+/// used to access a value identifier by such key the prefix of the key (e.g. `something`). Please also observe the usual
+/// Swift naming conventions, e.g. prefer `ID` to `Id` etc.
+///
+/// ## Usage
+/// Using a baggage container is fairly straight forward, as it boils down to using the prepared computed properties:
+///
+///     var context = Baggage.topLevel
+///     // set a new value
+///     context.testID = "abc"
```

Comment:
@ktoso I'm wondering if naming it `context` instead of `baggage` still makes sense in this case, as it's not the actual context object but its storage.

Reply:
good catch yes this should be baggage.